File: frontend/src/api/llmCatalog.ts
import { client } from './client';

export type LLMModelCatalogItem = {
  model_key: string;
  provider_key: string;
  display_name: string;
  enabled: boolean;
  sort_order: number;
  llm_metadata_json: Record<string, unknown>;
};

export type LLMProviderCatalogItem = {
  provider_key: string;
  display_name: string;
  provider_type: 'hosted' | 'local';
  credential_provider?: string | null;
  enabled: boolean;
  sort_order: number;
  metadata_json: Record<string, unknown>;
  models: LLMModelCatalogItem[];
};

export type LLMCatalogResponse = {
  providers: LLMProviderCatalogItem[];
};

export async function getLlmCatalog(): Promise<LLMCatalogResponse> {
  const { data, error } = await client.GET('/api/llm-catalog' as never);

  if (error) {
    throw error;
  }

  return data as LLMCatalogResponse;
}

File: backend/api/schemas/llm_catalog.py
from typing import Literal

from pydantic import BaseModel, Field


class LLMModelResponse(BaseModel):
    model_key: str
    provider_key: str
    display_name: str
    enabled: bool
    sort_order: int
    llm_metadata_json: dict = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class LLMProviderResponse(BaseModel):
    provider_key: str
    display_name: str
    provider_type: Literal["hosted", "local"]
    credential_provider: str | None = None
    enabled: bool
    sort_order: int
    metadata_json: dict = Field(default_factory=dict)
    models: list[LLMModelResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class LLMCatalogResponse(BaseModel):
    providers: list[LLMProviderResponse] = Field(default_factory=list)
